sonar-batch/src/main/java/org/sonar/batch/cpd/CpdExecutor.java

```diff
@@ -17,45 +17,99 @@
 import org.sonar.batch.index.BatchComponent;
 import org.sonar.batch.index.BatchComponentCache;
 import org.sonar.batch.protocol.output.BatchReport;
 import org.sonar.batch.protocol.output.BatchReport.Duplicate;
 import org.sonar.batch.protocol.output.BatchReport.Duplication;
 import org.sonar.batch.report.ReportPublisher;
+import org.sonar.duplications.block.Block;
+import org.sonar.duplications.detector.suffixtree.SuffixTreeCloneDetectionAlgorithm;
 import org.sonar.duplications.index.CloneGroup;
 import org.sonar.duplications.index.ClonePart;
+import org.sonar.duplications.index.PackedMemoryCloneIndex.ResourceBlocks;
 
-import static com.google.common.collect.FluentIterable.from;
-
-public abstract class AbstractCpdEngine extends CpdEngine {
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
 
-  private static final Logger LOG = Loggers.get(AbstractCpdEngine.class);
+import static com.google.common.collect.FluentIterable.from;
 
+/**
+ * Runs on the root module, at the end of the project analysis.
+ * It executes copy paste detection involving all files of all modules, which were indexed during sensors execution for each module
+ * by {@link CpdSensor). The sensor is responsible for handling exclusions and block sizes.
+ */
+public class CpdExecutor {
+  private static final Logger LOG = Loggers.get(CpdExecutor.class);
   static final int MAX_CLONE_GROUP_PER_FILE = 100;
   static final int MAX_CLONE_PART_PER_GROUP = 100;
 
+  private final SonarDuplicationsIndex index;
   private final ReportPublisher publisher;
   private final BatchComponentCache batchComponentCache;
+  private final Settings settings;
 
-  public AbstractCpdEngine(ReportPublisher publisher, BatchComponentCache batchComponentCache) {
+  public CpdExecutor(Settings settings, SonarDuplicationsIndex index, ReportPublisher publisher, BatchComponentCache batchComponentCache) {
+    this.settings = settings;
+    this.index = index;
     this.publisher = publisher;
     this.batchComponentCache = batchComponentCache;
   }
 
-  protected final void saveDuplications(final InputFile inputFile, List<CloneGroup> duplications) {
+  public void execute() {
+    Iterator<ResourceBlocks> it = index.iterator();
+
+    while (it.hasNext()) {
+      ResourceBlocks resourceBlocks = it.next();
+      runCpdAnalysis(resourceBlocks.resourceId(), resourceBlocks.blocks());
+    }
+  }
+
+  private void runCpdAnalysis(String resource, Collection<Block> fileBlocks) {
+    LOG.debug("Detection of duplications for {}", resource);
+
+    List<CloneGroup> duplications;
+    try {
+      duplications = SuffixTreeCloneDetectionAlgorithm.detect(index, fileBlocks);
+    } catch (Exception e) {
+      throw new IllegalStateException("Fail during detection of duplication for " + resource, e);
+    }
+
+    BatchComponent component = batchComponentCache.get(resource);
+    InputFile inputFile = (InputFile) component.inputComponent();
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) NullPointerException might be thrown as 'component' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)